### PR TITLE
Better behavior when doing recursive filtering

### PR DESCRIPTION
**Pull request by aaschaer**
For https://app.shortcut.com/globus/story/17393/update-globus-ls-recursive-filter-x-to-use-new-transfer-filter-behavior-to-include-dirs

Depends on https://github.com/globus/globus-sdk-python/pull/652, so I'm making this a draft for now.

Also want to make sure we're all on the same page about expected behavior. Some example output:

```
$ globus ls $EP1:/ --recursive --filter ~*1.txt
home/
mnt/
not shareable/
share/
home/aaschaer/
home/aaschaer/file1.txt
not shareable/godata/
not shareable/godata/file1.txt
share/godata/
share/godata/file1.txt
```

### CODE REVIEW

File: changelog.d/20221117_103859_aaschaer_recursive_filter.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+### Enhancements
+
+* the `--filter` and `--recursive` options to `globus ls` now combine better
```
[the remark this answers is not in the document]
Comment (aaschaer) [reply changelog.d/20221117_103859_aaschaer_recursive_filter.md:3]: Yeah, "combine better" isn't great, still trying to think of a good way to explain this behavior

File: changelog.d/20221117_103859_aaschaer_recursive_filter.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+  with directories that don't match the name filter being explored
```
Comment (sirosen) [inline changelog.d/20221117_103859_aaschaer_recursive_filter.md:4]: May I suggest this longer, but hopefully more digestible, explanation?

```suggestion
* `globus ls` has improved behavior when the `--filter` and `--recursive` options
   are used in combination

  * directory names are not matched against the filter, allowing the operation to
    traverse directories regardless of their names

  * the `--filter` is still applied to filenames in all directories traversed by
    the `ls` operation

  * directory names can be filtered out of the text output by eliminating
    lines which end in `/`

  * the behaviors of `globus ls` commands with `--recursive` or `--filter`, but not
    both, are unchanged
```

I'm happy for this to get further refined, but I think we need to explain how they behave when combined here, and it's good to note that we aren't changing how they behave when _not_ combined.

File: src/globus_cli/commands/ls.py [hunk 3]
```diff
@@ -180,29 +180,32 @@ def ls_command(
...
-        # surprising
-        # if we're asked to change or "improve" the behavior in the future, we
-        # could do so with "type:dir" or "type:file" filters added in, and
-        # potentially work out some viable behavior based on what people want
+
+        # if we are doing filtering we need to pass multiple filter params. The
+        # first allows all directories, as we need them for recursive
+        # expansion. The second then filters name by the filter_val
+        if filter_val:
+            ls_params["filter"] = ["type:dir", f"name:{filter_val}"]
```
Comment (sirosen) [inline src/globus_cli/commands/ls.py:197]: If I recall our formatting code correctly, we can also leave the `:` formatting to the SDK, like so:
```suggestion
            ls_params["filter"] = [{"type": "dir"}, {"name": filter_val}]
```

Do we like this better? I think it's pretty much the same, but maybe others will find it easier to read or have a stronger opinion.

File: src/globus_cli/services/transfer/recursive_ls.py [hunk 4]
```diff
@@ -60,13 +58,11 @@ def __init__(
         ls_params: dict[str, t.Any],
         *,
         max_depth: int = 3,
-        filter_after_first: bool = True,
```
Comment (sirosen) [inline src/globus_cli/services/transfer/recursive_ls.py:63]: Was this param simply unused? I can't recall why we needed it, but it seems like maybe just a vestige (in which case, 👍 for removing it!).

**Review by sirosen — COMMENTED** [review]
I like that we're including the directory names in the output. I don't know if we discussed and agreed on that, but it's easier to ask people to filter them out (e.g. with `grep`) if they don't want them than it is to control in the CLI and give an option.

File: src/globus_cli/services/transfer/recursive_ls.py (hunk 4 above)
Comment (aaschaer) [reply src/globus_cli/services/transfer/recursive_ls.py:63]: Yeah, I was worried it was something I'd have to figure out how to support, but it was always passed as True, so I just removed it

File: src/globus_cli/commands/ls.py (hunk 3 above)
Comment (aaschaer) [reply src/globus_cli/commands/ls.py:197]: Oh, I like that a bit more, I'm just going to add it

**Comment by aaschaer** [comment]
> I like that we're including the directory names in the output. I don't know if we discussed and agreed on that, but it's easier to ask people to filter them out (e.g. with grep) if they don't want them than it is to control in the CLI and give an option.

That was one of the things I wanted to make sure we were all okay with. The original story suggests "And client-side, filter out directories (?)." Which I would be fine adding, but it wasn't clear if we would want to filter out all directories or just directories that didn't match the filter.

File: changelog.d/20221117_103859_aaschaer_recursive_filter.md (hunk 2 above)
Comment (aaschaer) [reply changelog.d/20221117_103859_aaschaer_recursive_filter.md:4]: This is significantly better, so I'm going to add the suggested commit, and then we can modify as needed

**Review by kurtmckee — APPROVED** [review]

**Pull request merged**